Libs/MRML/Core/vtkMRMLTableStorageNode.cxx

```diff
@@ -340,106 +340,273 @@ bool vtkMRMLTableStorageNode::ReadTable(std::string filename, vtkMRMLTableNode*
-      typedColumn->SetNumberOfTuples(numberOfTuples);
-
-      // Initialize with null value
-      std::string nullValueStr = tableNode->GetColumnProperty(columnName, "nullValue");
-      if (typedColumn->IsNumeric())
-        {
-        // numeric arrays can be initialized in one batch
-        double nullValue = 0.0;
-        if (!nullValueStr.empty())
-          {
-          nullValue = vtkVariant(nullValueStr).ToDouble();
-          }
-        typedColumn->FillComponent(0, nullValue);
-        }
-      else
-        {
-        vtkVariant nullValue(nullValueStr);
-        for (vtkIdType row = 0; row < numberOfTuples; ++row)
-          {
-          typedColumn->SetVariantValue(row, nullValue);
-          }
-        }
-
-      // Set values
-      if (valueTypeId == VTK_CHAR || valueTypeId == VTK_SIGNED_CHAR || valueTypeId == VTK_UNSIGNED_CHAR)
+      if (components.size() > 0)
         {
-        bool valid = false;
-        for (vtkIdType row = 0; row < numberOfTuples; ++row)
+        vtkAbstractArray* columnArray = components[0];
+        if (columnArray)
           {
-          if (column->GetValue(row).empty())
-            {
-            // empty cell, leave the null value
-            continue;
-            }
-          int value = column->GetVariantValue(row).ToInt(&valid);
-          if (!valid)
-            {
-            continue;
-            }
-          typedColumn->SetVariantValue(row, vtkVariant(value));
-          }
-        }
-      else
-        {
-        for (vtkIdType row = 0; row < numberOfTuples; ++row)
-          {
-          if (column->GetValue(row).empty())
-            {
-            // empty cell, leave the null value
-            continue;
-            }
-          typedColumn->SetVariantValue(row, column->GetVariantValue(row));
+          columnArray->SetName(columnName.c_str());
+          table->AddColumn(columnArray);
           }
         }
+      }
+     else
+       {
+       // Output column. Can be multi-component
+       vtkSmartPointer<vtkDataArray> typedColumn = vtkSmartPointer<vtkDataArray>::Take(vtkDataArray::CreateDataArray(valueTypeId));
+       typedColumn->SetName(columnName.c_str());
+       typedColumn->SetNumberOfComponents(components.size());
+       vtkIdType numberOfTuples = 0;
+       for (vtkAbstractArray* componentArray : components)
+         {
+         if (componentArray == nullptr)
+           {
+           continue;
+           }
+         numberOfTuples = std::max(numberOfTuples, componentArray->GetNumberOfTuples());
+         }
+       typedColumn->SetNumberOfTuples(numberOfTuples);
+
+       int componentIndex = 0;
+       for (vtkAbstractArray* componentColumn : components)
+         {
+         if (componentColumn == nullptr)
+           {
+           continue;
+           }
+
+         vtkStringArray* rawColumn = vtkStringArray::SafeDownCast(componentColumn);
+         if (!rawColumn)
+           {
+           continue;
```

Comment:
we should log a warning and maybe even insert a dummy column (with all empty values)